### ISSUE TITLE
Biome API: top_node variant if it's below dust_node

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#6386]
Suggestion for the Biome API:

In the biome definition, you have `top_node` for the top node and `dust_node` for snowfall.

Now I want to achieve the following:
I want to have a normal dirt-with-grass as a `top_node` and a snow (`default:snow`) as a `dust_node`. But now comes the trick: The `top_node` must change to dirt-with-snow when the `dust_node` has been placed on it.

This has the following nice effect: Layer snow is nicely on top of dirt-with snow under open sky. Under tree leaves, there will be no snow at all.

Long story short: I want to turn this:
![](https://i.imgur.com/3VJYxLT.png)

into this:
![](https://i.imgur.com/KGwHAqq.png)

Definition:

* `top_node` is the default top node to use, like now.
* If the new optional field `top_node_dust` is set, the top node becomes `top_node_dust` when a `dust_node` is placed on top by the mapgen.
* If `top_node_dust` is not set, all top nodes are `top_node` regardless of dust (this is current behaviour)

**Comment by tenplus1** [comment]
Use a schematic with -1 y placement to place snow but has dirt_with_snow below it.

**Comment by Wuzzy2** [comment]
I don't think this will work unless you can show me a schematic which actually does this.
I don't see how a schematic knows the height position of the floor, which might vary like in the screenshot.

**Comment by paramat** [comment]
I don't yet see enough useful use for this.

> This has the following nice effect: Layer snow is nicely on top of dirt-with snow under open sky. Under tree leaves, there will be no snow at all.

For this specific use i disagree, snow drifts sideways a lot so you get a thinner layer of snow under trees, which current behaviour does very nicely. The patches of green spoil the completely-white appearence of a snowy forest and are unrealistic.

**Comment by paramat** [comment]
You can't use a schematic for dust nodes.

**Comment by Wuzzy2** [comment]
Is this a good justification?:

![](https://i.imgur.com/yZh1q6p.jpg)

Currently, dirt-with-snow is always placed, even under massive terrain overhangs. Above example shows how it would look like if this is *not* the case. Here, all blocks below a terrain overhang are not snowed in.

**Comment by paramat** [comment]
> all blocks below a terrain overhang are not snowed in which would be the correct behaviour IMO.

No, i explained above that snow drifts sideways a lot as it falls and will cover ground that is not directly under sky.
I did a websearch of snowy pine forests and you don't get grass patches under trees, it really spoils the snowy forest look, and this looks wrong too.
Currently you get a thinner layer of snow under an overhang, which is realistic.
